### PR TITLE
Add support for storyboard samples

### LINKED ISSUE #54: iOS Safari throws error 4 on audio loading
Jeez, this browser is so dumb... I can't even find where this error is happening, because it just throws 4 as a number... Through trial and error, I figured out that it's something related to [Howler's onloaderror event](https://github.com/goldfire/howler.js#onloaderror-function). I have no idea how to fix this.

image

### PR DESCRIPTION
**Pull request by kionell**
This should add support for storyboard samples only. Hit sounds are more complex to handle, so I'm not sure if I can add them anytime soon.

Example beatmaps: 
https://osu.ppy.sh/beatmapsets/3756#osu/22538 - uses storyboard samples for voice lines.
https://osu.ppy.sh/beatmapsets/5381#osu/26547 - uses a whole bunch of samples for everything, including custom hit object sounds.

### CODE REVIEW

File: src/play/loader/sample-loader.ts [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+        let decoded = 0;
+        for (const [name, blob] of input.entries()) {
+          cb(
+            decoded / input.size,
+            `Loading Samples (${decoded + 1} / ${input.size})`
+          );
+
+          const howl = new Howl({
+            src: URL.createObjectURL(blob),
+            html5: true,
```
Comment (blueberrymuffin3) [inline src/play/loader/sample-loader.ts]: Is there any reason in particular that your using html5 audio here?

Comment (blueberrymuffin3) [reply src/play/loader/sample-loader.ts]: Using HTML5 audio exposes the underlying audio to the operating system, allowing things like pausing with hardware pause keys.

File: src/play/render/common/storyboard_sample.ts [hunk 2]
```diff
@@ -0,0 +1,16 @@
...
+  sound: Howl | null;
+
+  constructor(object: StoryboardSample, samples: Map<string, Howl>) {
+    this.sound = samples.get(object.filePath) ?? null;
+
+    if (!this.sound) return;
+
+    this.sound.once("end", () => this.sound!.unload());
+    this.sound.volume(object.volume / 100);
+    this.sound.play();
```
Comment (blueberrymuffin3) [inline src/play/render/common/storyboard_sample.ts]: Could you use howl's sound id's to allow the same sound to be played multiple times? You probably don't need to unload it either.

**Review by blueberrymuffin3 — CHANGES_REQUESTED** [review]

File: src/play/loader/sample-loader.ts (hunk 1 above)
Comment (kionell) [reply src/play/loader/sample-loader.ts]: No, i don't have any reasons for using it. I just left it as it was in the audio loading.

File: src/play/render/common/storyboard_sample.ts (hunk 2 above)
Comment (kionell) [reply src/play/render/common/storyboard_sample.ts]: > Could you use howl's sound id's

Some samples can be played multiple times simultaneously. If i understand correctly, you can only control one track using a sound id.

> You probably don't need to unload it either.

You are right, my bad.

Comment (kionell) [reply src/play/render/common/storyboard_sample.ts]: Modern browsers don't allow audio autoplay, which can delay sample playback when switching tabs.

**Comment by blueberrymuffin3** [comment]
I believe that the delay in audio when switching tabs is actually caused by the main event loop, which uses `requestAnimationFrame`, not being called. We can either schedule the sounds within the webaudio context, or just ignore the issue.

**Comment by kionell** [comment]
I still get these warnings. It's about Chrome [autoplay policy](https://developer.chrome.com/blog/autoplay/). Chrome suspends all audio that was played without user gestures on the background. I don't know if it's possible to get around this. 

image

**Comment by blueberrymuffin3** [comment]
> I still get these warnings. It's about Chrome [autoplay policy](https://developer.chrome.com/blog/autoplay/). Chrome suspends all audio that was played without user gestures on the background. I don't know if it's possible to get around this.
> 
> image

This isn't actually an issue as clicking the start button counts as a user gesture, and no sound is played before then.

**Review by blueberrymuffin3 — APPROVED** [review]

**Comment by blueberrymuffin3** [comment]
@kionell TYSM for your work!

**Pull request merged**